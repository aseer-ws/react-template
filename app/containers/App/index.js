/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import GlobalStyle from '@app/global-styles';
import { routeConfig } from '@app/routeConfig';
import For from '@components/For';
import Header from '@components/Header';
import { colors } from '@themes';
import { Layout } from 'antd';
import map from 'lodash/map';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { withRouter } from 'react-router';
import { Route, Switch } from 'react-router-dom';
import { compose } from 'redux';
import { ThemeProvider } from 'styled-components';

const theme = {
  fg: colors.primary,
  bg: colors.secondary
};

export function App({ location, history }) {
  useEffect(() => {
    if (location.search.includes('?redirect_uri=')) {
      const routeToReplace = new URLSearchParams(location.search).get('redirect_uri');
      history.replace(routeToReplace);
    }
  }, []);

  return (
    <ThemeProvider theme={theme}>
      <Header />
      <Layout.Content>
        <For
          ParentComponent={(props) => <Switch {...props} />}
          of={map(Object.keys(routeConfig))}
          renderItem={(routeKey, index) => {
            const Component = routeConfig[routeKey].component;
            return (
              <Route
                exact={routeConfig[routeKey].exact}
                key={index}
                path={routeConfig[routeKey].route}
                render={(props) => {
                  const updatedProps = {
                    ...props,
                    ...routeConfig[routeKey].props
                  };
                  return <Component {...updatedProps} />;
                }}
              />
            );
          }}
        />
        <GlobalStyle />
      </Layout.Content>
    </ThemeProvider>
  );
}
App.propTypes = {
  location: PropTypes.object,
  history: PropTypes.object
};
export default compose(withRouter)(App);
