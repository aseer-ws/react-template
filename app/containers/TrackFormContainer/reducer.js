import produce from 'immer';
import { createActions } from 'reduxsauce';

export const GENRE_ENUM = ['Jaz', 'HipHop', 'Electronics', 'Pop'];

export const initialValues = {
  trackName: '',
  trackPrice: 1,
  primaryGenreName: GENRE_ENUM[0],
  releaseDate: '22/01/2015',
  trackViewUrl: '',
  collectionName: '',
  collectionPrice: 2,
  collectionViewUrl: '',
  artistName: '',
  artistViewUrl: ''
};

export const fillValues = {
  trackName: 'Track A',
  trackPrice: 2,
  primaryGenreName: GENRE_ENUM[0],
  releaseDate: '02/11/2018',
  trackViewUrl: 'https://tracka.com',
  collectionName: 'Collection B',
  collectionPrice: 12,
  collectionViewUrl: 'https://collectionb.com',
  artistName: 'Artist C',
  artistViewUrl: 'https://artistc.com'
};

export const initialState = {
  formValues: initialValues
};

export const { Types: trackFormContainerTypes, Creators: trackFormContainerCreators } = createActions({
  setFormValues: ['formValues'],
  resetForm: {},
  fillForm: {}
});

const trackFormContainerReducer = function (state = initialState, action) {
  return produce(state, (draft) => {
    switch (action.type) {
      case trackFormContainerTypes.SET_FORM_VALUES:
        draft.formValues = { ...state.formValues, ...action.formValues };
        break;
      case trackFormContainerTypes.RESET_FORM:
        draft.formValues = initialValues;
        break;
      case trackFormContainerTypes.FILL_FORM:
        draft.formValues = fillValues;
        break;
      default:
        return initialState;
    }
  });
};

export default trackFormContainerReducer;
