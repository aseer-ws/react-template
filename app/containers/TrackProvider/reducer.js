/*
 *
 * TrackProvider reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const initialState = {
  list: { artistName: '', tracks: [], trackCount: 0, error: null },
  details: {
    trackId: null,
    track: {},
    error: null
  }
};

export const { Types: trackProviderTypes, Creators: trackProviderCreators } = createActions({
  requestGetTracks: ['artistName'],
  successGetTracks: ['data'],
  failureGetTracks: ['error'],
  clearTracks: {},
  requestGetTrack: ['trackId'],
  successGetTrack: ['data'],
  failureGetTrack: ['error']
});

/* eslint-disable default-case, no-param-reassign */
export const trackProviderReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackProviderTypes.REQUEST_GET_TRACKS:
        draft.list.artistName = action.artistName;
        break;
      case trackProviderTypes.SUCCESS_GET_TRACKS:
        draft.list.tracks = action.data.results;
        draft.list.trackCount = action.data.resultCount;
        draft.list.error = null;
        break;
      case trackProviderTypes.FAILURE_GET_TRACKS:
        draft.list.tracks = [];
        draft.list.trackCount = 0;
        draft.list.error = action.error;
        break;
      case trackProviderTypes.CLEAR_TRACKS:
        draft.list.tracks = [];
        draft.list.trackCount = 0;
        draft.list.error = null;
        break;
      case trackProviderTypes.REQUEST_GET_TRACK:
        draft.details.trackId = action.trackId;
        break;
      case trackProviderTypes.SUCCESS_GET_TRACK:
        draft.details.track = action.data;
        draft.details.error = null;
        break;
      case trackProviderTypes.FAILURE_GET_TRACK:
        draft.details.track = {};
        draft.details.error = action.error;
        break;
      default:
        return state;
    }
  });

export default trackProviderReducer;
