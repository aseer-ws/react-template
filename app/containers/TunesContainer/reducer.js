/*
 *
 * TunesContainer reducer
 *
 */
import produce from 'immer';
// import { fromJS } from 'immutable';
import { createActions } from 'reduxsauce';

export const initialState = {
  artist: '',
  songsData: {},
  tunesError: null
};

export const { Types: tunesContainerTypes, Creators: tunesContainerCreators } = createActions({
  setArtist: ['artistName'],
  setSongsData: ['songsData'],
  setTunesError: ['error'],
  clearSongs: {}
});

/* eslint-disable default-case, no-param-reassign */
export const tunesContainerReducer = produce((state = initialState, action) => {
  switch (action.type) {
    case tunesContainerTypes.SET_ARTIST:
      state.artist = action.artistName;
      break;
    case tunesContainerTypes.SET_SONGS_DATA:
      state.tunesError = null;
      state.songsData = action.songsData;
      break;
    case tunesContainerTypes.SET_TUNES_ERROR:
      state.songsData = {};
      state.tunesError = action.error;
      break;
    case tunesContainerTypes.CLEAR_SONGS:
      state.songsData = {};
      state.artist = '';
      break;
    default:
      return state;
  }
});

export default tunesContainerReducer;
