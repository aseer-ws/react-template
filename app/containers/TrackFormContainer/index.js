/**
 *
 * TrackFormContainer Container
 *
 */

import Container from '@app/components/Container';
import If from '@app/components/If';
import { T } from '@app/components/T';
import TrackModal from '@app/components/TrackFormModal';
import { submitTune } from '@app/services/tunesApi';
import history from '@app/utils/history';
import { Button, Card, Divider, Form, Input, message, notification, Space, Spin, Steps } from 'antd';
import moment from 'moment';
import PropTypes from 'prop-types';
import React, { memo, useState } from 'react';
import { Helmet } from 'react-helmet';
import { injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import styled from 'styled-components';
import { fillValues, initialValues, trackFormContainerCreators } from './reducer';
import { selectFormValues } from './selectors';
import trackFormSteps, {
  DATE_FORMAT,
  FORM_LOADING_TIP,
  STEP_ROUTES,
  STEP_TITLES,
  TOTAL_STEPS,
  trackFormProps,
  TRACK_SUBMIT_FAILED,
  TRACK_SUBMIT_SUCCESS
} from './trackFormDetails';
import useFormStepRedirect from './useFormStepRedirect';

const StyledHeader = styled.header`
  height: 5rem;
  display: grid;
  align-items: center;
  padding-bottom: '2rem';
`;

const StyledHeaderText = styled(T)`
  text-align: center;
`;

const SpreadSpace = styled(Space)`
  && {
    justify-content: center;
  }
`;

const CenteredStep = styled(Steps.Step)`
  && {
    .ant-steps-item-title {
      font-size: 0.65rem;
      font-weight: bold;
    }
    .ant-steps-icon {
      top: 50%;
      transform: translateY(-50%);
      display: grid;
      place-items: center;
    }
  }
`;

function getBase64(img, callback) {
  const reader = new FileReader();
  reader.addEventListener('load', () => callback(reader.result));
  reader.readAsDataURL(img);
}

export function TrackFormContainer({ maxWidth, formValues, setFormValues, resetForm, fillForm }) {
  const stepTitle = new URLSearchParams(history.location.search).get('stepTitle');
  const [form] = Form.useForm();
  const [loading, setLoading] = useState(false);
  const [showTrack, setShowTrack] = useState(false);
  const [loadingForm, setLoadingForm] = useState(stepTitle !== 'track');

  const displayForm = () => setLoadingForm(false);
  const toggleSpin = () => setLoading((load) => !load);

  useFormStepRedirect(stepTitle, formValues, displayForm);

  const moveToStep = (step) => {
    history.push(`/add-track?stepTitle=${STEP_ROUTES[step]}`);
  };

  const handleSubmit = async (values) => {
    toggleSpin();
    setFormValues(values);
    if (stepTitle === STEP_ROUTES[TOTAL_STEPS - 1]) {
      const res = await submitTune({ ...formValues, ...values });
      if (res.ok) {
        notification.success({ message: TRACK_SUBMIT_SUCCESS });
        setShowTrack(true);
      } else {
        notification.error({ message: TRACK_SUBMIT_FAILED, description: res.data });
      }
    } else {
      moveToStep(STEP_TITLES[stepTitle] + 1);
    }
    toggleSpin();
  };

  const handleReset = () => {
    resetForm();
    form.setFieldsValue({ ...initialValues, releaseDate: moment(initialValues.releaseDate, DATE_FORMAT) });
    moveToStep(0);
  };

  const handleFillForm = () => {
    fillForm();
    form.setFieldsValue({ ...fillValues, releaseDate: moment(fillValues.releaseDate, DATE_FORMAT) });
    moveToStep(TOTAL_STEPS - 1);
  };

  const handlePrev = () => {
    moveToStep(STEP_TITLES[stepTitle] - 1);
  };

  const onFileChange = function (info) {
    if (info.file.status === 'done') {
      getBase64(info.file.originFileObj, (imgUrl) => {
        form.setFieldsValue({ artworkUrl: imgUrl });
        message.success(`${info.file.name} file uploaded.`);
      });
    }
  };

  const formContainer = (
    <>
      <Steps
        current={STEP_TITLES[stepTitle]}
        status="process"
        data-testid="step-parent"
        style={{ marginBottom: '2rem' }}
      >
        {STEP_ROUTES.map((stepTitle, index) => (
          <CenteredStep data-testid={`${index}_step`} key={stepTitle} title={String(stepTitle).toUpperCase()} />
        ))}
      </Steps>
      <Form
        {...trackFormProps}
        form={form}
        initialValues={{ ...formValues, releaseDate: moment(formValues?.releaseDate, DATE_FORMAT) }}
        onFinish={handleSubmit}
      >
        {/* Step Form Inputs */}
        {trackFormSteps[stepTitle]?.map(({ itemProps, component: Component = Input, componentProps }) => (
          <Form.Item
            data-testid={`${itemProps.name}_test`}
            key={`loopStep_${itemProps.name}`}
            {...itemProps}
            {...(itemProps.name === 'artworkUrl' ? { getValueFromEvent: onFileChange } : {})}
          >
            <Component data-testid={itemProps.name} {...componentProps} />
          </Form.Item>
        ))}
        <SpreadSpace style={{ textAlign: 'center', width: '100%', margin: '1rem auto' }}>
          <If condition={loading}>
            <Spin />
            <T text={STEP_ROUTES[TOTAL_STEPS - 1] === stepTitle ? 'Submitting' : 'Validating'} />
          </If>
        </SpreadSpace>
        <Divider />
        {/* Form Footer/Navigator */}
        <Form.Item wrapperCol={{ span: 24 }}>
          <SpreadSpace direction="vertical" size={10} style={{ textAlign: 'center', width: '100%' }}>
            <Space>
              <Button
                data-testid="prev-btn"
                disabled={STEP_TITLES[stepTitle] === 0}
                htmlType="button"
                onClick={handlePrev}
              >
                Prev
              </Button>
              <Button data-testid="submit-btn" type="primary" htmlType="submit">
                {stepTitle === STEP_ROUTES[TOTAL_STEPS - 1] ? 'Submit' : 'Next'}
              </Button>
            </Space>
            {/* {process.env.NODE_ENV === 'development' && ( */}
            <Space>
              <>
                <Button data-testid="reset-btn" htmlType="button" onClick={handleReset}>
                  Reset
                </Button>
                <Button data-testid="fill-btn" type="dashed" htmlType="button" onClick={handleFillForm}>
                  Fill form
                </Button>
              </>
            </Space>
            {/* )} */}
          </SpreadSpace>
        </Form.Item>
      </Form>
    </>
  );

  return (
    <Container maxWidth={maxWidth}>
      <Helmet>
        <title>Add Track</title>
        <meta name="description" content="Add new Track" />
      </Helmet>
      <Card>
        <StyledHeader>
          <StyledHeaderText type="heading" id="add_track_header_text" />
        </StyledHeader>

        <Spin size="large" spinning={loadingForm} tip={FORM_LOADING_TIP}>
          {formContainer}
        </Spin>
        <TrackModal
          handleCancel={() => setShowTrack(false)}
          trackDetails={formValues}
          isVisible={showTrack}
          afterClose={handleReset}
        />
      </Card>
    </Container>
  );
}

TrackFormContainer.propTypes = {
  maxWidth: PropTypes.number.isRequired,
  formValues: PropTypes.object.isRequired,
  setFormValues: PropTypes.func.isRequired,
  resetForm: PropTypes.func.isRequired,
  fillForm: PropTypes.func.isRequired
};

const mapStateToProps = createStructuredSelector({
  formValues: selectFormValues()
});

const { setFormValues, fillForm, resetForm } = trackFormContainerCreators;

export function mapDispatchToProps(dispatch) {
  return {
    setFormValues: (fieldValues) => dispatch(setFormValues(fieldValues)),
    resetForm: () => dispatch(resetForm()),
    fillForm: () => dispatch(fillForm())
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, memo)(TrackFormContainer);

export const TrackFormContainerTest = compose(injectIntl)(TrackFormContainer);

// [DONE] move to next step on enter
// [DONE] remove step state. on single source of truth
// [DONE] submit next prev button mystery
// theming
// mobile experience
// accessibility
