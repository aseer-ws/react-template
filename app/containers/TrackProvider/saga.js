import { getTunes } from '@app/services/tunesApi';
import { call, put, select, takeLatest } from 'redux-saga/effects';
import { trackProviderCreators, trackProviderTypes } from './reducer';
import { selectCurrentTrackFromTracks } from './selectors';

const { successGetTracks, failureGetTracks, successGetTrack, failureGetTrack } = trackProviderCreators;

export function* getArtistTracks(action) {
  const response = yield call(getTunes, action.artistName);
  const { ok, data, problem } = response;
  if (ok && data) {
    yield put(successGetTracks(data));
  } else {
    yield put(failureGetTracks(data ?? problem));
  }
}

export const CURRENT_TRACK_FROM_TRACKS_SELECTOR = selectCurrentTrackFromTracks();

export function* getCurrentTrack(action) {
  const trackFromTrackList = yield select(CURRENT_TRACK_FROM_TRACKS_SELECTOR);
  if (!trackFromTrackList) {
    const res = yield call(getTunes, action.trackId);
    const { ok, data, problem } = res;
    if (ok && data && data?.results) {
      yield put(successGetTrack(data.results[0]));
    } else {
      yield put(failureGetTrack(data ?? problem));
    }
  } else {
    yield put(successGetTrack(trackFromTrackList));
  }
}

export default function* trackProviderSaga() {
  yield takeLatest(trackProviderTypes.REQUEST_GET_TRACK, getCurrentTrack);
  yield takeLatest(trackProviderTypes.REQUEST_GET_TRACKS, getArtistTracks);
}
