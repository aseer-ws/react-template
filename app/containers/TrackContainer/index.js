/**
 *
 * TrackContainer
 *
 */

import CollectionModal from '@app/components/CollectionModal';
import Container from '@app/components/Container';
import If from '@app/components/If';
import { getLimitLineCSS, T } from '@app/components/T';
import { colors, media } from '@app/themes';
import { Card, Col, Image, Row, Skeleton, Statistic } from 'antd';
import { isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import React, { memo, useEffect, useState } from 'react';
import { Helmet } from 'react-helmet';
import { injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import { useParams } from 'react-router-dom';
import { compose } from 'redux';
import { injectSaga } from 'redux-injectors';
import { createStructuredSelector } from 'reselect';
import styled from 'styled-components';
import { trackProviderCreators } from '@app/containers/TrackProvider/reducer';
import trackProviderSaga from '@app/containers/TrackProvider/saga';
import { selectTrack, selectTrackError } from '@app/containers/TrackProvider/selectors';

const TrackContent = styled.div`
  position: relative;
  width: 100%;
  display: flex;
  gap: 1rem;
  align-items: center;

  ${media.lessThan('tablet')`
    flex-direction: column !important;
  `}
`;

const StyledImageContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  width: 100%;
  flex: 0.4;
`;

const StyledImage = styled(Image)`
  && {
    height: 100%;
    width: 100%;
    object-fit: cover;
  }
`;

const StyledAudio = styled.audio`
  display: block;
  margin-top: 1rem;
`;

const TrackDetails = styled.div`
  flex: 0.6;
`;

const TrackName = styled(T)`
  && {
    color: ${colors.gotoStories};
    font-size: 1.6rem;
    margin-bottom: 0.5rem;
    font-weight: 700;
  }
`;
const ArtistName = styled(T)`
  && {
    margin-top: 0rem;
    font-size: 1.2rem;
    font-weight: 600;
    margin-bottom: 0.6rem;
    opacity: 0.5;
  }
`;

const ReleaseDate = styled(T)`
  && {
    margin-bottom: 0.2rem;
    font-weight: bold;
    color: #444;
    width: 100%;
  }
`;

const CollectionName = styled.p`
  && {
    margin: 0.5rem 0;
    cursor: pointer;
    ${getLimitLineCSS(3)}
    animation: blink infinite 1.5s;

    @keyframes blink {
      0% {
        color: ${colors.gotoStories};
      }
      50% {
        opacity: 0.5;
      }
      100% {
        color: ${colors.text};
      }
    }

    &:hover {
      color: ${colors.success};
    }
  }
`;

const TrackGenre = styled(T)`
  && {
    margin-top: 0.5rem;
    font-size: smaller;
    font-weight: bold;
    padding: 0.5rem;
    background-color: #d8fa43;
    width: max-content;
    border-radius: 999px;
  }
`;

const { requestGetTrack } = trackProviderCreators;

export function dateFormat(date) {
  return new Date(date).toLocaleDateString('en-IN', {
    month: 'short',
    day: 'numeric',
    year: 'numeric'
  });
}

const valueStyle = { fontSize: '1.2rem' };

export function TrackContainer({ dispatchGetTrack, track, maxWidth, padding, trackError }) {
  const [loading, setLoading] = useState(true);
  const params = useParams();

  useEffect(() => {
    if (!isEmpty(track) || !isEmpty(trackError)) {
      setLoading(false);
    }
    let loadingTimeout = setTimeout(() => {
      setLoading(false);
    }, 3000);
    return () => {
      clearTimeout(loadingTimeout);
    };
  }, [track, trackError]);

  useEffect(() => {
    if (dispatchGetTrack) {
      dispatchGetTrack(params.trackId);
    }
  }, [params.trackId]);

  return (
    <Container maxWidth={maxWidth} padding={padding}>
      <Helmet>
        <title>
          {track?.trackName
            ? `${track?.trackName}${track?.artistName && ` by ${track?.artistName}`}`
            : 'TrackContainer'}
        </title>
        <meta name="description" content={track?.trackName ?? 'Description of TrackContainer'} />
      </Helmet>
      <Skeleton data-testid="loader" loading={loading} active>
        <Card>
          <TrackContent>
            <StyledImageContainer>
              <StyledImage
                width="95%"
                height="95%"
                src={track?.artworkUrl100 ?? ''}
                fallback="https://media.istockphoto.com/vectors/thumbnail-image-vector-graphic-vector-id1147544807?k=20&m=1147544807&s=612x612&w=0&h=pBhz1dkwsCMq37Udtp9sfxbjaMl27JUapoyYpQm0anc="
              />
            </StyledImageContainer>
            <TrackDetails>
              <If
                condition={!isEmpty(track?.trackName)}
                otherwise={<TrackName data-testid="track_name_unavailable" id="track_name_unavailable" />}
              >
                <TrackName noOfLines={3} title={track?.trackName} data-testid="track-name" text={track?.trackName} />
              </If>
              <If
                condition={!isEmpty(track?.artistName)}
                otherwise={<ArtistName data-testid="artist_name_unavailable" id="artist_name_unavailable" />}
              >
                <ArtistName data-testid="artist-name" text={track?.artistName} />
              </If>
              <If
                condition={!isEmpty(track?.releaseDate)}
                otherwise={<ReleaseDate data-testid="release_date_unavailable" id="release_date_unavailable" />}
              >
                <ReleaseDate data-testid="release-date" text={`Release Date: ${dateFormat(track?.releaseDate)}`} />
              </If>
              <If
                condition={!isEmpty(track?.collectionName)}
                otherwise={<T data-testid="collection_name_unavailable" id="collection_name_unavailable" />}
              >
                <CollectionModal
                  collectionId={track?.collectionId}
                  collectionName={track?.collectionName}
                  collectionPrice={track?.collectionPrice}
                  collectionViewUrl={track?.collectionViewUrl}
                >
                  <CollectionName noOFLines={3} data-testid="collection-name">
                    {track?.collectionName}
                  </CollectionName>
                </CollectionModal>
              </If>
              <Row>
                <Col span={8}>
                  <Statistic
                    data-testid="track-price"
                    suffix="$"
                    precision={2}
                    title="Track Price"
                    value={track?.trackPrice}
                    valueStyle={valueStyle}
                  />
                </Col>
                <Col span={8}>
                  <Statistic
                    data-testid="track-time"
                    precision={2}
                    suffix=" s"
                    title="Track Time"
                    value={track?.trackTimeMillis / 1000}
                    valueStyle={valueStyle}
                  />
                </Col>
                <Col span={8}>
                  <Statistic
                    data-testid="track-country"
                    precision={2}
                    title="Country"
                    value={track?.country}
                    valueStyle={valueStyle}
                  />
                </Col>
              </Row>
              <If condition={typeof track?.trackPrice !== 'number'}>
                <T data-testid="track_price_unavailable" id="track_price_unavailable" />
              </If>

              <If
                condition={!isEmpty(track?.primaryGenreName)}
                otherwise={<TrackGenre data-testid="track_genre_unavailable" id="track_genre_unavailable" />}
              >
                <TrackGenre data-testid="track-genre" text={track?.primaryGenreName} />
              </If>
              <If condition={!isEmpty(track?.previewUrl)} otherwise={<T id="track_preview_unavailable" />}>
                <StyledAudio loading="lazy" preload="none" controls src={track?.previewUrl}>
                  Your browser does not support <code>audio</code> element
                </StyledAudio>
              </If>
            </TrackDetails>
          </TrackContent>
        </Card>
      </Skeleton>
    </Container>
  );
}

TrackContainer.propTypes = {
  dispatchGetTrack: PropTypes.func,
  track: PropTypes.object,
  trackError: PropTypes.any,
  maxWidth: PropTypes.number,
  padding: PropTypes.number
};

const mapStateToProps = createStructuredSelector({
  track: selectTrack(),
  trackError: selectTrackError()
});

export function mapDispatchToProps(dispatch) {
  return {
    dispatchGetTrack: (trackId) => dispatch(requestGetTrack(trackId))
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  withConnect,
  memo,
  injectSaga({ key: 'trackProvider', saga: trackProviderSaga })
)(TrackContainer);

export const TrackContainerTest = compose(injectIntl)(TrackContainer);
