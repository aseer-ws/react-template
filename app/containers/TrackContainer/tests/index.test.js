/**
 *
 * Tests for TrackContainer
 *
 *
 */

import { translate } from '@app/components/IntlGlobalProvider';
import { dateFormat, mapDispatchToProps } from '@app/containers/TrackContainer';
import { trackProviderTypes } from '@app/containers/TrackProvider/reducer';
import { renderProvider, timeout } from '@utils/testUtils';
import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { TrackContainerTest as TrackContainer } from '../index';

describe('<TrackContainer /> container tests', () => {
  let submitSpy;
  let trackInfo = {
    artistId: 466532,
    collectionId: 1165630592,
    trackId: 1165631048,
    artistName: 'Korn',
    collectionName: 'Follow The Leader',
    trackName: 'Justin',
    artistViewUrl: 'https://music.apple.com/us/artist/korn/466532?uo=4',
    collectionViewUrl: 'https://music.apple.com/us/album/justin/1165630592?i=1165631048&uo=4',
    trackViewUrl: 'https://music.apple.com/us/album/justin/1165630592?i=1165631048&uo=4',
    previewUrl:
      'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/54/57/70/545770b4-fa22-3d84-50fd-8007f97a4192/mzaf_8941792383747410636.plus.aac.p.m4a',
    artworkUrl100:
      'https://is3-ssl.mzstatic.com/image/thumb/Music125/v4/3d/9c/6b/3d9c6bdb-6572-a68d-dd5c-400c51531a71/source/100x100bb.jpg',
    trackPrice: 1.29,
    releaseDate: '1998-08-18T12:00:00Z',
    currency: 'USD',
    primaryGenreName: 'Hard Rock'
  };

  beforeEach(() => {
    submitSpy = jest.fn();
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<TrackContainer />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render song info', async () => {
    const { getByTestId } = renderProvider(<TrackContainer track={{ ...trackInfo }} dispatchGetTrack={submitSpy} />);
    await timeout(200);
    expect(getByTestId('track-name')).toHaveTextContent(trackInfo.trackName);
    expect(getByTestId('artist-name')).toHaveTextContent(trackInfo.artistName);
    expect(getByTestId('collection-name')).toHaveTextContent(trackInfo.collectionName);
    expect(getByTestId('track-price')).toHaveTextContent(trackInfo.trackPrice);
    expect(getByTestId('track-genre')).toHaveTextContent(trackInfo.primaryGenreName);
    expect(getByTestId('release-date')).toHaveTextContent(dateFormat(trackInfo.releaseDate));
  });

  it('should render fallback message if no data provided', async () => {
    const { getByTestId } = renderProvider(<TrackContainer dispatchGetTrack={submitSpy} />);
    // wait for skeleton to dispappear
    await timeout(4000);
    expect(getByTestId('track_name_unavailable')).toHaveTextContent(translate('track_name_unavailable'));
    expect(getByTestId('artist_name_unavailable')).toHaveTextContent(translate('artist_name_unavailable'));
    expect(getByTestId('collection_name_unavailable')).toHaveTextContent(translate('collection_name_unavailable'));
    expect(getByTestId('track_price_unavailable')).toHaveTextContent(translate('track_price_unavailable'));
    expect(getByTestId('release_date_unavailable')).toHaveTextContent(translate('release_date_unavailable'));
    expect(getByTestId('track_genre_unavailable')).toHaveTextContent(translate('track_genre_unavailable'));
  });

  it('should mapDispatchToProps works as expected', () => {
    const props = mapDispatchToProps(submitSpy);

    const actions = {
      dispatchGetTrack: {
        type: trackProviderTypes.REQUEST_GET_TRACK,
        trackId: 123
      }
    };
    props.dispatchGetTrack(123);
    expect(submitSpy).toHaveBeenCalledWith(actions.dispatchGetTrack);
  });
});
