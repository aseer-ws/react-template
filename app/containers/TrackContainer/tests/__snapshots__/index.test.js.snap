// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackContainer /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="Container-sc-1vnn72p-0 iOkZIh"
    >
      <div
        class="ant-skeleton ant-skeleton-active"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;
